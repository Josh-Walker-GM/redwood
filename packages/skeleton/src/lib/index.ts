import fs from 'fs'

import { parse } from '@babel/parser'

export function getFileAST(filePath: string) {
  const code = fs.readFileSync(filePath, { encoding: 'utf8', flag: 'r' })
  return parse(code, {
    sourceType: 'unambiguous',
    plugins: ['jsx', 'typescript'],
  })
}
