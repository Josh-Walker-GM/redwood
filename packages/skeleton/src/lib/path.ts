import path from 'path'

import { getBaseDir, getConfigPath } from '@redwoodjs/internal/dist/index'

/**
 * Determines the root path of a redwood project which contains the provided path
 * @param pathWithinProject A path within a redwood project
 * @returns The root path of the associated project
 */
export function getRootPath(pathWithinProject: string) {
  return pathWithinProject
    ? path.dirname(getConfigPath(pathWithinProject))
    : getBaseDir()
}
