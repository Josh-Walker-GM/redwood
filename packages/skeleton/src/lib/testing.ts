import { ensurePosixPath } from '@redwoodjs/internal/dist/paths'

export function stripAndFormatPathForTesting(
  filepath: string,
  projectFilepath: string
) {
  if (filepath.startsWith(projectFilepath)) {
    filepath = ensurePosixPath(filepath.substring(projectFilepath.length))
  }
  return filepath
}
