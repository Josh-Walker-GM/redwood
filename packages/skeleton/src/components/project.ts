import fs from 'fs'
import path from 'path'

import chalk from 'chalk'

import { getPaths } from '@redwoodjs/internal/dist/index'

import { getRootPath } from '../lib/path'

import { RedwoodSkeleton } from './base'
import { extractCells, RedwoodCell } from './cell'
import { extractFunctions, RedwoodFunction } from './function'
import { extractLayouts, RedwoodLayout } from './layout'
import { extractPages, RedwoodPage } from './page'
import { extractRouters, RedwoodRouter } from './router'
import { extractSides, RedwoodSide } from './side'

/**
 * Used to enumerate either JS or TS project types
 */
export enum RedwoodProjectType {
  TYPESCRIPT = 'typescript',
  JAVASCRIPT = 'javascript',
}

export class RedwoodProject extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  readonly type: RedwoodProjectType

  private sides?: RedwoodSide[] | undefined
  private cells?: RedwoodCell[] | undefined
  private routers?: RedwoodRouter[] | undefined
  private layouts?: RedwoodLayout[] | undefined
  private pages?: RedwoodPage[] | undefined
  private functions?: RedwoodFunction[] | undefined

  public static getProject({
    pathWithinProject = '',
    full = false,
    readFromCache = true,
    insertIntoCache = true,
  }: {
    pathWithinProject?: string
    full?: boolean
    readFromCache?: boolean
    insertIntoCache?: boolean
  } = {}) {
    const projectCache = RedwoodProjectsCache.getInstance().projects
    const rootPath = getRootPath(pathWithinProject)

    if (readFromCache) {
      const cachedProject = projectCache.get(rootPath)
      if (cachedProject) {
        return cachedProject
      }
    }

    const project = new RedwoodProject(rootPath, full)
    if (insertIntoCache) {
      projectCache.set(project.filepath, project)
    }
    return project
  }

  private constructor(rootPath: string, full = false) {
    super(rootPath)

    // A project is typescript if we detect a tsconfig.json
    this.type =
      fs.existsSync(
        path.join(getPaths(this.filepath).web.base, 'tsconfig.json')
      ) ||
      fs.existsSync(
        path.join(getPaths(this.filepath).api.base, 'tsconfig.json')
      )
        ? RedwoodProjectType.TYPESCRIPT
        : RedwoodProjectType.JAVASCRIPT

    if (full) {
      this.sides = extractSides(this)
      this.cells = extractCells(this)
      this.layouts = extractLayouts(this)
      this.pages = extractPages(this)
      this.functions = extractFunctions(this)
      this.routers = extractRouters(this)
    }
  }

  getSides(forceExtract = false): RedwoodSide[] {
    if (forceExtract) {
      return extractSides(this)
    }
    return this.sides === undefined ? extractSides(this) : this.sides
  }

  getCells(forceExtract = false): RedwoodCell[] {
    if (forceExtract) {
      return extractCells(this)
    }
    return this.cells === undefined ? extractCells(this) : this.cells
  }

  getRouters(forceExtract = false): RedwoodRouter[] {
    if (forceExtract) {
      return extractRouters(this)
    }
    return this.routers === undefined ? extractRouters(this) : this.routers
  }

  getLayouts(forceExtract = false): RedwoodLayout[] {
    if (forceExtract) {
      return extractLayouts(this)
    }
    return this.layouts === undefined ? extractLayouts(this) : this.layouts
  }

  getPages(forceExtract = false): RedwoodPage[] {
    if (forceExtract) {
      return extractPages(this)
    }
    return this.pages === undefined ? extractPages(this) : this.pages
  }

  getFunctions(forceExtract = false): RedwoodFunction[] {
    if (forceExtract) {
      return extractFunctions(this)
    }
    return this.functions === undefined
      ? extractFunctions(this)
      : this.functions
  }

  // Diagnostics

  getInformation(): string {
    return '' // TODO: Implement, gets the info of all other components too
  }

  printInformation(): void {
    console.log(
      `${chalk.bgCyan('[Info]')}\t${this.name}\t${chalk.dim(this.filepath)}`
    )
  }

  printWarnings(cascade = false): void {
    if (this.warnings.length > 0) {
      const titleLine = `${chalk.bgYellow('[Warn]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const warningLines = this.warnings.map((warning, index) => {
        return ` (${index + 1}) ${warning}\n`
      })
      console.log(titleLine.concat('\n', ...warningLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printWarnings()
      })
      this.functions?.forEach((func) => {
        func.printWarnings()
      })
      this.layouts?.forEach((layout) => {
        layout.printWarnings()
      })
      this.pages?.forEach((page) => {
        page.printWarnings()
      })
      this.routers?.forEach((router) => {
        router.printWarnings()
        router.routes.forEach((route) => {
          route.printWarnings()
        })
      })
      this.sides?.forEach((side) => {
        side.printWarnings()
      })
    }
  }

  printErrors(cascade = false): void {
    if (this.errors.length > 0) {
      const titleLine = `${chalk.bgRed('[Error]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const errorLines = this.errors.map((error, index) => {
        return ` (${index + 1}) ${error}\n`
      })
      console.log(titleLine.concat('\n', ...errorLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printErrors()
      })
      this.functions?.forEach((func) => {
        func.printErrors()
      })
      this.layouts?.forEach((layout) => {
        layout.printErrors()
      })
      this.pages?.forEach((page) => {
        page.printErrors()
      })
      this.routers?.forEach((router) => {
        router.printErrors()
        router.routes.forEach((route) => {
          route.printErrors()
        })
      })
      this.sides?.forEach((side) => {
        side.printErrors()
      })
    }
  }
}

/**
 * Singelton class which contains a cache of {@link RedwoodProject} instances.
 *
 * This class is not exported and should only be available from static methods within {@link RedwoodProject}, in order to keep the surface of the skeleton API cleaner and leaner.
 */
class RedwoodProjectsCache {
  private static instance: RedwoodProjectsCache

  public projects: Map<string, RedwoodProject>

  private constructor() {
    this.projects = new Map()
  }

  public static getInstance(): RedwoodProjectsCache {
    if (!RedwoodProjectsCache.instance) {
      RedwoodProjectsCache.instance = new RedwoodProjectsCache()
    }
    return RedwoodProjectsCache.instance
  }
}
