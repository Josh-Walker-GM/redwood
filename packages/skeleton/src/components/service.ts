import fs from 'fs'
import path from 'path'

import { getPaths } from '@redwoodjs/internal/dist/paths'

// import { getASTFromFile } from '../lib/ast'

import { RedwoodError, RedwoodWarning } from './diagnostic'
import type { RedwoodProject } from './project'
import { RedwoodSkeleton } from './skeleton'

export class RedwoodService extends RedwoodSkeleton {
  warnings: RedwoodWarning[] = []
  errors: RedwoodError[] = []

  constructor(filepath: string) {
    super(filepath)

    // const _ = getASTFromFile(filepath)

    // TODO: Extract all functions
    // TODO: Extract all function parameters

    // Checks

    // TODO: Check service has corresponding SDL
    // TODO: Check function parameters match corresponding query or mutation parameters
  }
}

export function extractService(filepath: string): RedwoodService {
  return new RedwoodService(filepath)
}

export function extractServices(
  project: RedwoodProject | undefined = undefined
): RedwoodService[] {
  const services: RedwoodService[] = []

  const servicesPath = project
    ? getPaths(project.filepath).api.services
    : getPaths().api.services

  if (!fs.existsSync(servicesPath)) {
    return services
  }

  // TODO: Confirm this is the condition to detect a service
  const getServiceFiles = (directory: string) => {
    const serviceFiles: string[] = []
    const directoryContents = fs.readdirSync(directory)
    directoryContents.forEach((content) => {
      const stat = fs.lstatSync(path.join(directory, content))
      if (stat.isDirectory()) {
        serviceFiles.push(...getServiceFiles(path.join(directory, content)))
      } else if (stat.isFile()) {
        if (
          content.match(/.+\.(js|ts)$/) &&
          !content.match(/.+\.(test\.js|test\.ts)$/) &&
          !content.match(/.+\.(scenarios\.js|scenarios\.ts)$/)
        ) {
          serviceFiles.push(path.join(directory, content))
        }
      }
    })
    return serviceFiles
  }

  const serviceFiles = getServiceFiles(servicesPath)
  serviceFiles.forEach((servicePath) => {
    services.push(extractService(servicePath))
  })

  return services
}
