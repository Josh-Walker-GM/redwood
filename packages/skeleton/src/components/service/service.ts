import fs from 'fs'
import path from 'path'

import traverse from '@babel/traverse'

import { getPaths } from '@redwoodjs/internal/dist/paths'

import { getASTFromFile } from '../../lib/ast'
import type { RedwoodProject } from '../project'
import { RedwoodSkeleton } from '../skeleton'

import { RedwoodServiceFunction } from './function'

export class RedwoodService extends RedwoodSkeleton {
  readonly functions: RedwoodServiceFunction[] = []

  constructor(filepath: string) {
    super(filepath)
    const ast = getASTFromFile(filepath)
    traverse(ast, {
      ExportNamedDeclaration: (path) => {
        this.functions.push(
          new RedwoodServiceFunction(this.filepath, path.node)
        )
      },
    })
  }

  executeAdditionalChecks() {
    this.functions.forEach((func) => {
      func.executeAdditionalChecks()
    })
  }
}

export function extractService(filepath: string) {
  return new RedwoodService(filepath)
}

export function extractServices(project?: RedwoodProject) {
  const services: RedwoodService[] = []

  const servicesPath = project
    ? getPaths(project.filepath).api.services
    : getPaths().api.services

  if (!fs.existsSync(servicesPath)) {
    return services
  }

  // TODO: Confirm this is the condition to detect a service
  const getServiceFiles = (directory: string) => {
    const serviceFiles: string[] = []
    const directoryContents = fs.readdirSync(directory)
    directoryContents.forEach((content) => {
      const stat = fs.lstatSync(path.join(directory, content))
      if (stat.isDirectory()) {
        serviceFiles.push(...getServiceFiles(path.join(directory, content)))
      } else if (stat.isFile()) {
        if (
          content.match(/.+\.(js|ts)$/) &&
          !content.match(/.+\.(test\.js|test\.ts)$/) &&
          !content.match(/.+\.(scenarios\.js|scenarios\.ts)$/)
        ) {
          serviceFiles.push(path.join(directory, content))
        }
      }
    })
    return serviceFiles
  }

  const serviceFiles = getServiceFiles(servicesPath)
  serviceFiles.forEach((servicePath) => {
    services.push(extractService(servicePath))
  })

  return services
}
