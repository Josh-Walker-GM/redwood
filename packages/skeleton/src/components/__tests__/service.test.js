import path from 'path'

import { stripAndFormatPathForTesting } from '../../lib/testing'
import { RedwoodProject } from '../project'
import { extractServices } from '../service/service'

const FIXTURE_PATH = path.join(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '..',
  '__fixtures__'
)

describe.each([
  'empty-project',
  'example-todo-main',
  'example-todo-main-with-errors',
  'test-project',
])('From within the %s fixture', (PROJECT_NAME) => {
  const PROJECT_PATH = path.join(FIXTURE_PATH, PROJECT_NAME)

  const RWJS_CWD = process.env.RWJS_CWD
  beforeAll(() => {
    process.env.RWJS_CWD = PROJECT_PATH
  })
  afterAll(() => {
    process.env.RWJS_CWD = RWJS_CWD
  })

  it('returns the correct services without a project', () => {
    const services = extractServices(undefined)
    services.forEach((service) => {
      service.executeAdditionalChecks()
      service.filepath = stripAndFormatPathForTesting(
        service.filepath,
        PROJECT_PATH
      )
      service.functions.forEach((func) => {
        func.filepath = stripAndFormatPathForTesting(
          func.filepath,
          PROJECT_PATH
        )
      })
      expect(service).toMatchSnapshot(service.filepath)
    })
    expect(services.length).toMatchSnapshot('service count')
  })

  it('returns the correct services with a project', () => {
    const project = RedwoodProject.getProject({
      pathWithinProject: PROJECT_PATH,
      readFromCache: false,
      insertIntoCache: false,
    })
    const services = extractServices(project)
    services.forEach((service) => {
      service.executeAdditionalChecks()
      service.filepath = stripAndFormatPathForTesting(
        service.filepath,
        PROJECT_PATH
      )
      service.functions.forEach((func) => {
        func.filepath = stripAndFormatPathForTesting(
          func.filepath,
          PROJECT_PATH
        )
      })
      expect(service).toMatchSnapshot(service.filepath)
    })
    expect(services.length).toMatchSnapshot('service count')
  })
})
