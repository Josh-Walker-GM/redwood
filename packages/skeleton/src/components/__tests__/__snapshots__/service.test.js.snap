// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the empty-project fixture returns the correct services with a project: service count 1`] = `0`;

exports[`From within the empty-project fixture returns the correct services without a project: service count 1`] = `0`;

exports[`From within the example-todo-main fixture returns the correct services with a project: /api/src/services/todos/todos.js 1`] = `
RedwoodService {
  "errors": [],
  "filepath": "/api/src/services/todos/todos.js",
  "functions": [
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todos",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "createTodo",
      "parameters": [
        "body",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "numTodos",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "updateTodoStatus",
      "parameters": [
        "id",
        "status",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "renameTodo",
      "parameters": [
        "id",
        "body",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
  ],
  "name": "todos",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct services with a project: service count 1`] = `1`;

exports[`From within the example-todo-main fixture returns the correct services without a project: /api/src/services/todos/todos.js 1`] = `
RedwoodService {
  "errors": [],
  "filepath": "/api/src/services/todos/todos.js",
  "functions": [
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todos",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "createTodo",
      "parameters": [
        "body",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "numTodos",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "updateTodoStatus",
      "parameters": [
        "id",
        "status",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "renameTodo",
      "parameters": [
        "id",
        "body",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
  ],
  "name": "todos",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct services without a project: service count 1`] = `1`;

exports[`From within the example-todo-main-with-errors fixture returns the correct services with a project: /api/src/services/todos/todos.js 1`] = `
RedwoodService {
  "errors": [],
  "filepath": "/api/src/services/todos/todos.js",
  "functions": [
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todos",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithStringRole",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMultipleRoles",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithInvalidRole",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithInvalidRoles",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMissingRoles",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithBuiltInDeprecatedDirective",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMissingRolesAttribute",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMissingRolesAttributeNumeric",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "createTodo",
      "parameters": [
        "body",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "updateTodoStatus",
      "parameters": [
        "id",
        "status",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "renameTodo",
      "parameters": [
        "id",
        "body",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "deleteTodo",
      "parameters": [
        "id",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
  ],
  "name": "todos",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct services with a project: service count 1`] = `1`;

exports[`From within the example-todo-main-with-errors fixture returns the correct services without a project: /api/src/services/todos/todos.js 1`] = `
RedwoodService {
  "errors": [],
  "filepath": "/api/src/services/todos/todos.js",
  "functions": [
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todos",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithStringRole",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMultipleRoles",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithInvalidRole",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithInvalidRoles",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMissingRoles",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithBuiltInDeprecatedDirective",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMissingRolesAttribute",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "todosWithMissingRolesAttributeNumeric",
      "parameters": [],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "createTodo",
      "parameters": [
        "body",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "updateTodoStatus",
      "parameters": [
        "id",
        "status",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "renameTodo",
      "parameters": [
        "id",
        "body",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/todos/todos.js",
      "name": "deleteTodo",
      "parameters": [
        "id",
      ],
      "serviceName": "todos",
      "warnings": [],
    },
  ],
  "name": "todos",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct services without a project: service count 1`] = `1`;

exports[`From within the test-project fixture returns the correct services with a project: /api/src/services/contacts/contacts.ts 1`] = `
RedwoodService {
  "errors": [],
  "filepath": "/api/src/services/contacts/contacts.ts",
  "functions": [
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "contacts",
      "parameters": [],
      "serviceName": "contacts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "contact",
      "parameters": [
        "id",
      ],
      "serviceName": "contacts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "createContact",
      "parameters": [
        "input",
      ],
      "serviceName": "contacts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "updateContact",
      "parameters": [
        "id",
        "input",
      ],
      "serviceName": "contacts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "deleteContact",
      "parameters": [
        "id",
      ],
      "serviceName": "contacts",
      "warnings": [],
    },
  ],
  "name": "contacts",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct services with a project: /api/src/services/posts/posts.ts 1`] = `
RedwoodService {
  "errors": [],
  "filepath": "/api/src/services/posts/posts.ts",
  "functions": [
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "posts",
      "parameters": [],
      "serviceName": "posts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "post",
      "parameters": [
        "id",
      ],
      "serviceName": "posts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "createPost",
      "parameters": [
        "input",
      ],
      "serviceName": "posts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "updatePost",
      "parameters": [
        "id",
        "input",
      ],
      "serviceName": "posts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "deletePost",
      "parameters": [
        "id",
      ],
      "serviceName": "posts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "Post",
      "parameters": [],
      "serviceName": "posts",
      "warnings": [],
    },
  ],
  "name": "posts",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct services with a project: /api/src/services/users/users.ts 1`] = `
RedwoodService {
  "errors": [],
  "filepath": "/api/src/services/users/users.ts",
  "functions": [
    RedwoodServiceFunction {
      "errors": [
        {
          "code": "00.001",
          "message": "Could not determine the name of a named export",
        },
      ],
      "filepath": "/api/src/services/users/users.ts",
      "name": "users",
      "parameters": [],
      "serviceName": "users",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/users/users.ts",
      "name": "user",
      "parameters": [
        "id",
      ],
      "serviceName": "users",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/users/users.ts",
      "name": "User",
      "parameters": [],
      "serviceName": "users",
      "warnings": [],
    },
  ],
  "name": "users",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct services with a project: service count 1`] = `3`;

exports[`From within the test-project fixture returns the correct services without a project: /api/src/services/contacts/contacts.ts 1`] = `
RedwoodService {
  "errors": [],
  "filepath": "/api/src/services/contacts/contacts.ts",
  "functions": [
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "contacts",
      "parameters": [],
      "serviceName": "contacts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "contact",
      "parameters": [
        "id",
      ],
      "serviceName": "contacts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "createContact",
      "parameters": [
        "input",
      ],
      "serviceName": "contacts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "updateContact",
      "parameters": [
        "id",
        "input",
      ],
      "serviceName": "contacts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/contacts/contacts.ts",
      "name": "deleteContact",
      "parameters": [
        "id",
      ],
      "serviceName": "contacts",
      "warnings": [],
    },
  ],
  "name": "contacts",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct services without a project: /api/src/services/posts/posts.ts 1`] = `
RedwoodService {
  "errors": [],
  "filepath": "/api/src/services/posts/posts.ts",
  "functions": [
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "posts",
      "parameters": [],
      "serviceName": "posts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "post",
      "parameters": [
        "id",
      ],
      "serviceName": "posts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "createPost",
      "parameters": [
        "input",
      ],
      "serviceName": "posts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "updatePost",
      "parameters": [
        "id",
        "input",
      ],
      "serviceName": "posts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "deletePost",
      "parameters": [
        "id",
      ],
      "serviceName": "posts",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/posts/posts.ts",
      "name": "Post",
      "parameters": [],
      "serviceName": "posts",
      "warnings": [],
    },
  ],
  "name": "posts",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct services without a project: /api/src/services/users/users.ts 1`] = `
RedwoodService {
  "errors": [],
  "filepath": "/api/src/services/users/users.ts",
  "functions": [
    RedwoodServiceFunction {
      "errors": [
        {
          "code": "00.001",
          "message": "Could not determine the name of a named export",
        },
      ],
      "filepath": "/api/src/services/users/users.ts",
      "name": "users",
      "parameters": [],
      "serviceName": "users",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/users/users.ts",
      "name": "user",
      "parameters": [
        "id",
      ],
      "serviceName": "users",
      "warnings": [],
    },
    RedwoodServiceFunction {
      "errors": [],
      "filepath": "/api/src/services/users/users.ts",
      "name": "User",
      "parameters": [],
      "serviceName": "users",
      "warnings": [],
    },
  ],
  "name": "users",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct services without a project: service count 1`] = `3`;
